File: Samples/BarcodeScanner/js/js/scenario2_MultipleScanners.js
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

(function() {
    "use strict";

    // Types
    var BarcodeSymbologies = Windows.Devices.PointOfService.BarcodeSymbologies;

    // Page elements
    var scenarioStartScanningInstance1;
    var scenaioEndScanningInstance1;
    var scenarioStartScanningInstance2;
    var scenaioEndScanningInstance2;

    var retain1;
    var scanDataType1;
    var scanData1;
    var scanDataFormatted1;
    var retain2;
    var scanDataType2;
    var scanData2;
    var scanDataFormatted2;

    var page = WinJS.UI.Pages.define("/html/scenario2_MultipleScanners.html", {

        // On ready - set button states and add event listeners to the buttons.
        ready: function(element, options) {

            WinJS.log("Click a start scanning button to begin.", "sample", "status");

            retain1 = document.getElementById("retain1");
            scanDataType1 = document.getElementById("scanDataType1");
            scanData1 = document.getElementById("scanData1");
            scanDataFormatted1 = document.getElementById("scanDataFormatted1");
            retain2 = document.getElementById("retain2");
            scanDataType2 = document.getElementById("scanDataType2");
            scanData2 = document.getElementById("scanData2");
            scanDataFormatted2 = document.getElementById("scanDataFormatted2");

            scenarioStartScanningInstance1 = document.getElementById("scenarioStartScanningInstance1");
            scenarioStartScanningInstance1.addEventListener("click", buttonStartScanningInstance1, false);
            scenarioStartScanningInstance1.disabled = false;

            scenarioEndScanningInstance1 = document.getElementById("scenarioEndScanningInstance1");
            scenarioEndScanningInstance1.addEventListener("click", buttonEndScanningInstance1, false);
            scenarioEndScanningInstance1.disabled = true;

            scenarioStartScanningInstance2 = document.getElementById("scenarioStartScanningInstance2");
            scenarioStartScanningInstance2.addEventListener("click", buttonStartScanningInstance2, false);
            scenarioStartScanningInstance2.disabled = false;

            scenarioEndScanningInstance2 = document.getElementById("scenarioEndScanningInstance2");
            scenarioEndScanningInstance2.addEventListener("click", buttonEndScanningInstance2, false);
            scenarioEndScanningInstance2.disabled = true;
        },

        unload: function() {
            closeScanner1();
            closeScanner2();
        }
    });

    var _activeBarcodeScannerInstance = { instance1: 0, instance2: 1 };
    var _scanner1 = null;
    var _claimedScanner1 = null;
    var _scanner2 = null;
    var _claimedScanner2 = null;

    var _currentInstance;

    // Create scanner instance 1 and claim and enable it.
    function buttonStartScanningInstance1() {

        // Create the barcode scanner. 
        createScanner(_activeBarcodeScannerInstance.instance1).then(function(success) {
            if (success) {
                // Claim the scanner and enable
                return claimAndEnableScanner(_activeBarcodeScannerInstance.instance1);
            }
        }).done(function(success) {
            if (success) {
                resetUI();
                scenarioStartScanningInstance1.disabled = true;
                scenarioEndScanningInstance1.disabled = false;
                scenarioStartScanningInstance2.disabled = false;
                scenarioEndScanningInstance2.disabled = true;
                _currentInstance = _activeBarcodeScannerInstance.instance1;
            }
        });
    }

    // Create scanner instance 2 and claim and enable it.
    function buttonStartScanningInstance2() {

        // Create the barcode scanner. 
        createScanner(_activeBarcodeScannerInstance.instance2).then(function(success) {
            if (success) {
                // Claim the scanner and enable
                return claimAndEnableScanner(_activeBarcodeScannerInstance.instance2);
            }
        }).done(function(success) {
            if (success) {
                resetUI();
                scenarioStartScanningInstance2.disabled = true;
                scenarioEndScanningInstance2.disabled = false;
                scenarioStartScanningInstance1.disabled = false;
                scenarioEndScanningInstance1.disabled = true;
                _currentInstance = _activeBarcodeScannerInstance.instance2;
            }
        });
    }

    // Reset UI after ending instance 1
    function buttonEndScanningInstance1() {
        closeScanner1();
        resetUI();
        scenarioStartScanningInstance1.disabled = false;
        scenarioEndScanningInstance1.disabled = true;
        WinJS.log("Click a start scanning button to begin.", "sample", "status");
    }

    // Reset UI after ending instance 2
    function buttonEndScanningInstance2() {
        closeScanner2();
        resetUI();
        scenarioStartScanningInstance2.disabled = false;
        scenarioEndScanningInstance2.disabled = true;
        WinJS.log("Click a start scanning button to begin.", "sample", "status");
    }

    // Resets the textContent in the scanned data values.
    function resetUI() {
        switch (_currentInstance) {
            case _activeBarcodeScannerInstance.instance1:
                scanDataType1.textContent = "No Data";
                scanData1.textContent = "No Data";
                scanDataFormatted1.textContent = "No Data";
                break;

            case _activeBarcodeScannerInstance.instance2:
                scanDataType2.textContent = "No Data";
                scanData2.textContent = "No Data";
                scanDataFormatted2.textContent = "No Data";
                break;
        }
    }

    // This method is called upon when a claim request is made on instance 1. If a retain request was placed on the device it rejects the new claim.
    function onReleasedeviceRequested1(args) {
        // Check if the instance wants to retain the device
        if (retain1.checked) {
            // Retain the device
            _claimedScanner1.retainDevice();
        }
    }

    // This method is called upon when a claim request is made on instance 2. If a retain request was placed on the device it rejects the new claim.
    function onReleasedeviceRequested2(args) {
        // Check if the instance wants to retain the device
        if (retain2.checked) {
            // Retain the device
            _claimedScanner2.retainDevice();
        }
    }

    // Claims and enables the scanner for the given scanner instance and returns true if success.
    function claimAndEnableScanner(instance) {
        switch (instance) {
            case _activeBarcodeScannerInstance.instance1:
                if (_scanner1) {
                    // Claim the scanner
                    return _scanner1.claimScannerAsync().then(function(claimedScanner) {
                        if (claimedScanner) {
                            _claimedScanner1 = claimedScanner;
                            claimedScanner.isDecodeDataEnabled = true;

                            // Add the event handlers
                            claimedScanner.addEventListener("datareceived", onDataReceived);
                            claimedScanner.addEventListener("releasedevicerequested", onReleasedeviceRequested1);

                            // Enable the Scanner
                            return claimedScanner.enableAsync().then(function() {
                                WinJS.log("Instance 1 ready to scan. Device ID: " + _claimedScanner1.deviceId, "sample", "status");
                                return true;
                            });
                        } else {
                            WinJS.log("Instance 1 claim barcode scanner failed.", "sample", "error");
                            return false;
                        }
                    });
                }
                break;
            case _activeBarcodeScannerInstance.instance2:
                if (_scanner2) {
                    // Claim the scanner
                    return _scanner2.claimScannerAsync().then(function(claimedScanner) {
                        if (claimedScanner) {
                            _claimedScanner2 = claimedScanner;
                            claimedScanner.isDecodeDataEnabled = true;

                            // Add the event handlers
                            claimedScanner.addEventListener("datareceived", onDataReceived);
                            claimedScanner.addEventListener("releasedevicerequested", onReleasedeviceRequested2);

                            // Enable the Scanner
                            return claimedScanner.enableAsync().then(function() {
                                WinJS.log("Instance 2 ready to scan. Device ID: " + _claimedScanner2.deviceId, "sample", "status");
                                return true;
                            });
                        } else {
                            WinJS.log("Instance 2 claim barcode scanner failed.", "sample", "error");
                            return false;
                        }
                    });
                }
                break;
        }
    }

    function createScanner(instance) {
        return SdkSample.getFirstBarcodeScannerAsync().done(function (scanner) {
            if (scanner) {
                switch (instance) {
                    case _activeBarcodeScannerInstance.instance1:
                        _scanner1 = scanner;
                        break;
                    case _activeBarcodeScannerInstance.instance2:
                        _scanner2 = scanner;
                        break;
                }
                return true;
            } else {
                WinJS.log("Scanner not found. Please connect a barcode scanner.", "sample", "error");
                return false;
            }
        });
    }

    function closeScanner1() {
        if (_claimedScanner1) {
            // Remove event listeners and close.
            _claimedScanner1.removeEventListener("datareceived", onDataReceived);
            _claimedScanner1.removeEventListener("releasedevicerequested", onReleasedeviceRequested1);
            _claimedScanner1.close();
            _claimedScanner1 = null;
        }

        if (_scanner1) {
            _scanner1.close();
            _scanner1 = null;
        }
    }

    function closeScanner2() {
        if (_claimedScanner2) {
            // Remove event listeners and close.
            _claimedScanner2.removeEventListener("datareceived", onDataReceived);
            _claimedScanner2.removeEventListener("releasedevicerequested", onReleasedeviceRequested2);
            _claimedScanner2.close();
            _claimedScanner2 = null;
        }

        if (_scanner2) {
            _scanner2.close();
            _scanner2 = null;
        }
    }

    // This is an event handler for the claimed scanner Instance when it scans and recieves data
    function onDataReceived(args) {
        // Now populate the UI
        switch (_currentInstance) {
            case _activeBarcodeScannerInstance.instance1:
                scanDataType1.textContent = BarcodeSymbologies.getName(args.report.scanDataType);
                scanDataFormatted1.textContent = getDataLabelString(args.report.scanDataLabel, args.report.scanDataType);
                scanData1.textContent = getDataString(args.report.scanData);
                WinJS.log("Instance 1 received data from the barcode scanner.", "sample", "status");
                break;

            case _activeBarcodeScannerInstance.instance2:
                scanDataType2.textContent = BarcodeSymbologies.getName(args.report.scanDataType);
                scanDataFormatted2.textContent = getDataLabelString(args.report.scanDataLabel, args.report.scanDataType);
                scanData2.textContent = getDataString(args.report.scanData);
                WinJS.log("Instance 2 received data from the barcode scanner.", "sample", "status");
                break;
        }
    }
})();

File: Samples/BarcodeScanner/cppwinrt/Scenario2_MultipleScanners.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"

#include "Scenario2_MultipleScanners.h"
#include "Scenario2_MultipleScanners.g.cpp"

namespace winrt
{
    using namespace Windows::Devices;
    using namespace Windows::Devices::Enumeration;
    using namespace Windows::Devices::PointOfService;
    using namespace Windows::Foundation;
    using namespace Windows::UI;
    using namespace Windows::UI::Xaml;
    using namespace Windows::UI::Xaml::Controls;
    using namespace Windows::UI::Xaml::Media;
    using namespace Windows::UI::Xaml::Navigation;
}

namespace winrt::SDKTemplate::implementation
{

    Scenario2_MultipleScanners::Scenario2_MultipleScanners()
    {
        InitializeComponent();
    }

    /// <summary>
    /// Invoked when this page is no longer displayed.
    /// </summary>
    /// <param name="e">Event data that describes how this page was exited.  The Parameter
    /// property is typically used to configure the page.</param>
    void Scenario2_MultipleScanners::OnNavigatedFrom(NavigationEventArgs const&)
    {
        ResetTheScenarioState();
    }

    /// <summary>
    /// Invoked when this page is about to be displayed in a Frame.
    /// </summary>
    /// <param name="e">Event data that describes how this page was reached.  The Parameter
    /// property is typically used to configure the page.</param>
    void Scenario2_MultipleScanners::OnNavigatedTo(NavigationEventArgs const&)
    {
        ResetTheScenarioState();
    }

    /// <summary>
    /// This is the click handler for the 'ScenarioStartScanningInstance1' button. It initiates creation of scanner instance 1.
    /// </summary>
    fire_and_forget Scenario2_MultipleScanners::ButtonStartScanningInstance1_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        //Get the handle to the default scanner
        if (co_await CreateDefaultScannerObjectAsync(BarcodeScannerInstance::Instance1))
        {
            //Claim the scanner
            if (co_await ClaimBarcodeScannerAsync(BarcodeScannerInstance::Instance1))
            {
                //add the event handlers
                m_dev1ReleaseRequestedToken = m_claimedBarcodeScannerInstance1.ReleaseDeviceRequested({ get_weak(), &Scenario2_MultipleScanners::claimedBarcodeScannerInstance1_ReleaseDeviceRequested });
                m_dev1DataReceivedToken = m_claimedBarcodeScannerInstance1.DataReceived({ get_weak(), &Scenario2_MultipleScanners::claimedBarcodeScannerInstance1_DataReceived });
                m_claimedBarcodeScannerInstance1.IsDecodeDataEnabled(true);

                //Enable the Scanner
                if (co_await EnableBarcodeScannerAsync(BarcodeScannerInstance::Instance1))
                {
                    //Set the UI state
                    ResetUI();
                    SetUI(BarcodeScannerInstance::Instance1);
                }
            }
            else
            {
                if (m_barcodeScannerInstance1)
                {
                    m_barcodeScannerInstance1.Close();
                    m_barcodeScannerInstance1 = nullptr;
                }
            }
        }
    }

    /// <summary>
    /// This method is called upon when a claim request is made on instance 1. If a retain request was placed on the device it rejects the new claim.
    /// </summary>
    fire_and_forget Scenario2_MultipleScanners::claimedBarcodeScannerInstance1_ReleaseDeviceRequested(IInspectable const&, ClaimedBarcodeScanner const&)
    {
        auto lifetime = get_strong();

        co_await resume_foreground(Dispatcher());

        //check if the instance wants to retain the device
        if (Retain1().IsChecked().Value())
        {
            try
            {
                //Retain the device
                m_claimedBarcodeScannerInstance1.RetainDevice();
            }
            catch (hresult_error const& exception)
            {
                m_rootPage.NotifyUser(L"Retain instance 1 failed: " + exception.message(), NotifyType::ErrorMessage);
            }
        }
        //Release the device
        else
        {
            m_claimedBarcodeScannerInstance1.Close();
            m_claimedBarcodeScannerInstance1 = nullptr;

            if (m_barcodeScannerInstance1)
            {
                m_barcodeScannerInstance1.Close();
                m_barcodeScannerInstance1 = nullptr;
            }
        }
    }

    /// <summary>
    /// This is the click handler for the 'ScenarioStartScanningInstance2' button. Initiates creation of scanner instance 2
    /// </summary>
    fire_and_forget Scenario2_MultipleScanners::ButtonStartScanningInstance2_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        //Get the handle to the default scanner
        if (co_await CreateDefaultScannerObjectAsync(BarcodeScannerInstance::Instance2))
        {
            //Claim the scanner
            if (co_await ClaimBarcodeScannerAsync(BarcodeScannerInstance::Instance2))
            {
                //set the handlers
                m_dev2ReleaseRequestedToken = m_claimedBarcodeScannerInstance2.ReleaseDeviceRequested({ get_weak(), &Scenario2_MultipleScanners::claimedBarcodeScannerInstance2_ReleaseDeviceRequested });
                m_dev2DataReceivedToken = m_claimedBarcodeScannerInstance2.DataReceived({ get_weak(), &Scenario2_MultipleScanners::claimedBarcodeScannerInstance2_DataReceived });

                //enable the scanner to decode the scanned data
                m_claimedBarcodeScannerInstance2.IsDecodeDataEnabled(true);

                //Enable the Scanner
                if (co_await EnableBarcodeScannerAsync(BarcodeScannerInstance::Instance2))
                {
                    //Set the UI state
                    ResetUI();
                    SetUI(BarcodeScannerInstance::Instance2);
                }
            }
            else
            {
                if (m_barcodeScannerInstance2)
                {
                    m_barcodeScannerInstance2.Close();
                    m_barcodeScannerInstance2 = nullptr;
                }
            }
        }
    }

    /// <summary>
    /// This method is called upon when a claim request is made on instance 2. If a retain request was placed on the device it rejects the new claim.
    /// </summary>
    fire_and_forget Scenario2_MultipleScanners::claimedBarcodeScannerInstance2_ReleaseDeviceRequested(IInspectable const&, ClaimedBarcodeScanner const&)
    {
        auto lifetime = get_strong();

        co_await resume_foreground(Dispatcher());

        //check if the instance wants to retain the device
        if (Retain2().IsChecked().Value())
        {
            try
            {
                //Retain the device
                m_claimedBarcodeScannerInstance2.RetainDevice();
            }
            catch (hresult_error const& exception)
            {
                m_rootPage.NotifyUser(L"Retain instance 2 failed: " + exception.message(), NotifyType::ErrorMessage);
            }
        }
        //Release the device
        else
        {
            m_claimedBarcodeScannerInstance2.Close();
            m_claimedBarcodeScannerInstance2 = nullptr;

            if (m_barcodeScannerInstance2)
            {
                m_barcodeScannerInstance2.Close();
                m_barcodeScannerInstance2 = nullptr;
            }
        }
    }

    /// <summary>
    /// This is the click handler for the 'ScenarioEndScanningInstance1' button.
    /// Initiates the disposal of scanner instance 1.
    /// </summary>
    void Scenario2_MultipleScanners::ButtonEndScanningInstance1_Click(IInspectable const&, RoutedEventArgs const&)
    {
        if (m_claimedBarcodeScannerInstance1)
        {
            //remove the event handlers
            m_claimedBarcodeScannerInstance1.ReleaseDeviceRequested(m_dev1ReleaseRequestedToken);
            m_claimedBarcodeScannerInstance1.DataReceived(m_dev1DataReceivedToken);

            //dispose the instance
            m_claimedBarcodeScannerInstance1.Close();
            m_claimedBarcodeScannerInstance1 = nullptr;
        }

        if (m_barcodeScannerInstance1)
        {
            m_barcodeScannerInstance1.Close();
            m_barcodeScannerInstance1 = nullptr;
        }

        //reset the UI

        ResetUI();

        m_rootPage.NotifyUser(L"Click a start scanning button to begin.", NotifyType::StatusMessage);
    }

    /// <summary>
    /// This is the click handler for the 'ScenarioEndScanningInstance2' button.
    /// Initiates the disposal fo scanner instance 2.
    /// </summary>
    void Scenario2_MultipleScanners::ButtonEndScanningInstance2_Click(IInspectable const&, RoutedEventArgs const&)
    {
        if (m_claimedBarcodeScannerInstance2)
        {
            //remove the event handlers
            m_claimedBarcodeScannerInstance2.ReleaseDeviceRequested(m_dev2ReleaseRequestedToken);
            m_claimedBarcodeScannerInstance2.DataReceived(m_dev2DataReceivedToken);

            //dispose the instance
            m_claimedBarcodeScannerInstance2.Close();
            m_claimedBarcodeScannerInstance2 = nullptr;
        }

        if (m_barcodeScannerInstance2)
        {
            m_barcodeScannerInstance2.Close();
            m_barcodeScannerInstance2 = nullptr;
        }

        //reset the UI
        ResetUI();

        m_rootPage.NotifyUser(L"Click a start scanning button to begin.", NotifyType::StatusMessage);
    }

    /// <summary>
    /// This method returns the first available Barcode Scanner. To enumerate and find a particular device use the device enumeration code.
    /// </summary>
    /// <returns>a boolean value based on whether it found a compatible scanner connected</returns>

    IAsyncOperation<bool> Scenario2_MultipleScanners::CreateDefaultScannerObjectAsync(BarcodeScannerInstance instance)
    {
        auto lifetime = get_strong();

        BarcodeScanner scanner = co_await BarcodeScanner::GetDefaultAsync();

        if (!scanner)
        {
            m_rootPage.NotifyUser(L"Barcode scanner not found. Please connect a barcode scanner.", NotifyType::ErrorMessage);
            co_return false;
        }

        switch (instance)
        {
        case BarcodeScannerInstance::Instance1:
            m_barcodeScannerInstance1 = scanner;
            break;
        case BarcodeScannerInstance::Instance2:
            m_barcodeScannerInstance2 = scanner;
            break;
        default:
            co_return false;
        }

        co_return true;
    }

    /// <summary>
    /// This method claims the connected scanner.
    /// </summary>
    /// <returns>a boolean based on whether it was able to claim the scanner.</returns>

    IAsyncOperation<bool> Scenario2_MultipleScanners::ClaimBarcodeScannerAsync(BarcodeScannerInstance instance)
    {
        auto lifetime = get_strong();
        bool claimAsyncStatus = false;
        //select the instance to claim
        switch (instance)
        {
        case BarcodeScannerInstance::Instance1:

            m_claimedBarcodeScannerInstance1 = co_await m_barcodeScannerInstance1.ClaimScannerAsync();
            if (!m_claimedBarcodeScannerInstance1)
                m_rootPage.NotifyUser(L"Instance 1 claim barcode scanner failed.", NotifyType::ErrorMessage);
            else
                claimAsyncStatus = true;
            break;

        case BarcodeScannerInstance::Instance2:

            m_claimedBarcodeScannerInstance2 = co_await m_barcodeScannerInstance2.ClaimScannerAsync();
            if (!m_claimedBarcodeScannerInstance2)
                m_rootPage.NotifyUser(L"Instance 2 claim barcode scanner failed.", NotifyType::ErrorMessage);
            else
                claimAsyncStatus = true;
            break;

        default:
            co_return claimAsyncStatus;
        }
        co_return claimAsyncStatus;
    }

    /// <summary>
    /// This method enables the connected scanner.
    /// </summary>
    /// <returns>a boolean based on whether it was able to enable the scanner.</returns>

    IAsyncOperation<bool> Scenario2_MultipleScanners::EnableBarcodeScannerAsync(BarcodeScannerInstance instance)
    {
        auto lifetime = get_strong();
        switch (instance)
        {
        case BarcodeScannerInstance::Instance1:
            co_await m_claimedBarcodeScannerInstance1.EnableAsync();
            m_rootPage.NotifyUser(L"Instance 1 ready to scan. Device ID: " + m_claimedBarcodeScannerInstance1.DeviceId(), NotifyType::StatusMessage);
            break;
        case BarcodeScannerInstance::Instance2:
            co_await m_claimedBarcodeScannerInstance2.EnableAsync();
            m_rootPage.NotifyUser(L"Instance 2 ready to scan. Device ID: " + m_claimedBarcodeScannerInstance2.DeviceId(), NotifyType::StatusMessage);
            break;
        default:
            co_return false;
        }
        co_return true;
    }

    /// <summary>
    /// Reset the Scenario state
    /// </summary>

    void Scenario2_MultipleScanners::ResetTheScenarioState()
    {
        if (m_claimedBarcodeScannerInstance1)
        {
            m_claimedBarcodeScannerInstance1.Close();
            m_claimedBarcodeScannerInstance1 = nullptr;
        }

        if (m_barcodeScannerInstance1)
        {
            m_barcodeScannerInstance1.Close();
            m_barcodeScannerInstance1 = nullptr;
        }

        if (m_claimedBarcodeScannerInstance2)
        {
            m_claimedBarcodeScannerInstance2.Close();
            m_claimedBarcodeScannerInstance2 = nullptr;
        }

        if (m_barcodeScannerInstance2)
        {
            m_barcodeScannerInstance2.Close();
            m_barcodeScannerInstance2 = nullptr;
        }

        ResetUI();
    }

    /// <summary>
    /// Resets the display Elements to original state
    /// </summary>

    void Scenario2_MultipleScanners::ResetUI()
    {
        Instance1Border().BorderBrush(SolidColorBrush{ Colors::Gray() });
        Instance2Border().BorderBrush(SolidColorBrush{ Colors::Gray() });

        ScanDataType1().Text(L"No data");
        ScanData1().Text(L"No data");
        DataLabel1().Text(L"No data");
        ScanDataType2().Text(L"No data");
        ScanData2().Text(L"No data");
        DataLabel2().Text(L"No data");

        ScenarioStartScanningInstance1().IsEnabled(true);
        ScenarioStartScanningInstance2().IsEnabled(true);
        ScenarioEndScanningInstance1().IsEnabled(false);
        ScenarioEndScanningInstance2().IsEnabled(false);
    }

    /// <summary>
    /// Sets the UI elements to a state corresponding to the current active Instance.
    /// </summary>
    /// <param name="instance">Corresponds to the current active instance</param>

    void Scenario2_MultipleScanners::SetUI(BarcodeScannerInstance instance)
    {
        Instance1Border().BorderBrush(SolidColorBrush{ Colors::Gray() });
        Instance2Border().BorderBrush(SolidColorBrush{ Colors::Gray() });

        switch (instance)
        {
        case BarcodeScannerInstance::Instance1:
            ScenarioStartScanningInstance1().IsEnabled(false);
            ScenarioStartScanningInstance2().IsEnabled(true);
            ScenarioEndScanningInstance1().IsEnabled(true);
            ScenarioEndScanningInstance2().IsEnabled(false);
            Instance1Border().BorderBrush(SolidColorBrush{ Colors::DarkBlue() });
            break;

        case BarcodeScannerInstance::Instance2:
            ScenarioStartScanningInstance1().IsEnabled(true);
            ScenarioStartScanningInstance2().IsEnabled(false);
            ScenarioEndScanningInstance1().IsEnabled(false);
            ScenarioEndScanningInstance2().IsEnabled(true);
            Instance2Border().BorderBrush(SolidColorBrush{ Colors::DarkBlue() });
            break;
        }
    }

    /// <summary>
    /// This is an event handler for the claimed scanner Instance 1 when it scans and recieves data
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="args"></param>

    fire_and_forget Scenario2_MultipleScanners::claimedBarcodeScannerInstance1_DataReceived(ClaimedBarcodeScanner const&, BarcodeScannerDataReceivedEventArgs args)
    {
        auto lifetime = get_strong();
        co_await resume_foreground(Dispatcher());

        ScanDataType1().Text(BarcodeSymbologies::GetName(args.Report().ScanDataType()));
        DataLabel1().Text(GetDataLabelString(args.Report().ScanDataLabel(), args.Report().ScanDataType()));
        ScanData1().Text(GetDataString(args.Report().ScanData()));

        m_rootPage.NotifyUser(L"Instance 1 received data from the barcode scanner.", NotifyType::StatusMessage);
    }

    /// <summary>
    /// This is an event handler for the claimed scanner Instance 2 when it scans and recieves data
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="args"></param>

    fire_and_forget Scenario2_MultipleScanners::claimedBarcodeScannerInstance2_DataReceived(ClaimedBarcodeScanner const&, BarcodeScannerDataReceivedEventArgs args)
    {
        auto lifetime = get_strong();
        co_await resume_foreground(Dispatcher());
        ScanDataType2().Text(BarcodeSymbologies::GetName(args.Report().ScanDataType()));
        DataLabel2().Text(GetDataLabelString(args.Report().ScanDataLabel(), args.Report().ScanDataType()));
        ScanData2().Text(GetDataString(args.Report().ScanData()));

        m_rootPage.NotifyUser(L"Instance 2 received data from the barcode scanner.", NotifyType::StatusMessage);
    }

}